caffe2/opt/backend_cutting.cc

```diff
@@ -0,0 +1,416 @@
+
+namespace caffe2 {
+namespace opt {
+
+namespace {
+
+using NodeRef = nom::repr::NNGraph::NodeRef;
+using EdgeRef = nom::repr::NNGraph::EdgeRef;
+
+class GroupAnnotation {
+ public:
+  GroupAnnotation(int i, int g = -1) : group(g), in_degree(i) {}
+  int group;
+  int in_degree;
+  bool needs_transform{true};
+};
+
+struct VisitorContext {
+  VisitorContext(std::function<bool(const caffe2::OperatorDef&)> func)
+      : predicate(func) {}
+  std::unordered_map<NodeRef, GroupAnnotation> infos;
+  std::unordered_set<NodeRef> frontier;
+  std::vector<NodeRef> current_group;
+  std::function<bool(const caffe2::OperatorDef&)> predicate;
+
+  int group{0};
+  bool find_supported{true};
+};
+
+std::string ShowNode(NodeRef node) {
+  auto* node_data = node->data().get();
+  if (isa<nom::repr::NeuralNetData>(node_data)) {
+    const auto* nn_tensor = dyn_cast<const nom::repr::NeuralNetData>(node_data);
+    return MakeString("Tensor: ", nn_tensor->getName());
+  } else if (isa<nom::repr::NeuralNetOperator>(node_data)) {
+    const auto* nn_op = dyn_cast<const nom::repr::NeuralNetOperator>(node_data);
+    const auto* op_def = reinterpret_cast<const caffe2::OperatorDef*>(
+        nn_op->getAnnotation()->getSaved());
+    CAFFE_ENFORCE(op_def);
+    return MakeString("Op: ", op_def->type());
+  } else {
+    CAFFE_THROW("Known node");
+  }
+}
+
+void DumpGraph(nom::repr::NNGraph* g) {
+  auto nnprinter = [](typename nom::repr::NNGraph::NodeRef node) {
+    std::map<std::string, std::string> labelMap;
+    assert(node->data() && "Node doesn't have data, can't render it");
+    if (isa<nom::repr::NeuralNetOperator>(node->data())) {
+      auto* op = dyn_cast<nom::repr::NeuralNetOperator>(node->data().get());
+      labelMap["label"] =
+          op->getName() + " (" + std::to_string((unsigned long long)node) + ")";
+      auto* annotation = op->getAnnotation();
+      if (annotation && isa<nom::repr::DeviceAnnotation>(annotation)) {
+        auto device_annotation =
+            dyn_cast<nom::repr::DeviceAnnotation>(annotation);
+        labelMap["label"] += "\\n[" + device_annotation->getDevice() + "]";
+        auto hash = std::hash<std::string>{}(device_annotation->getDevice());
+        std::stringstream hex_stream;
+        hex_stream << std::hex << hash;
+        labelMap["color"] = "#" + hex_stream.str().substr(0, 6);
+        labelMap["fontcolor"] = labelMap["color"];
+      }
+      labelMap["shape"] = "box";
+    } else if (isa<nom::repr::Data>(node->data())) {
+      auto tensor = dyn_cast<nom::repr::NeuralNetData>(node->data().get());
+      labelMap["label"] = tensor->getName();
+      labelMap["label"] += "_" + std::to_string(tensor->getVersion()) + " " +
+          std::to_string((unsigned long long)node);
+    }
+    return labelMap;
+  };
+
+  std::ofstream out("dump.dot");
+  out << nom::converters::convertToDotString(g, nnprinter);
+  out.close();
+}
+
+// Explore the graph in topological order until we hit stopping nodes. This is
```

Comment:
might be nice to eventually add this to `Graph/Algorithms.h` in a slightly more generic fashion

Reply:
if you want to deal with loops, I'd recommend using `algorithm::tarjans`, which returns a reverse topologically sorted list of SCCs (which, in the trivial case of no cycles, all have a single element)

Reply:
Yeah, I was thinking about moving it to Algorithm as Kahn's algorithm is pretty generic but then I have this stopping condition...